tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/caches/Cache.scala

```diff
@@ -0,0 +1,170 @@
+package at.forsyte.apalache.tla.bmcmt.stratifiedRules.aux.caches
+
+import at.forsyte.apalache.tla.bmcmt.smt.SolverContext
+import at.forsyte.apalache.tla.bmcmt.stratifiedRules.aux.DelayedConstraintGenerator
+
+import scala.collection.immutable.HashMap
+
+/**
+ * Creates a stack-like bidirectional cache, which behaves in the following way: Suppose there exists a deterministic
+ * injective function `f: (ContextT, SourceT) -> (ContextT, TargetT)`, that is, a function which, given a value of type
+ * `SourceT`, computes a value of type `TargetT`, while mutating a context of type `ContextT`.
+ *
+ * [[Cache]] defines an entry-point `getOrCreate`, such that
+ *   - `getOrCreate(c, x) = f(c,x)`, if this is the first time `x` appears as the source value for `getOrCreate`.
+ *   - `getOrCreate(c, x) = (c, v)`, if `getOrCreate(c', x)` was previously called, and returned `(c'', v)` (for some
+ *     `c', c''`).
```

Comment:
Is the "mutation" of `ContextT` in place? If so, why do we need to return a reference to it? If it is instead a functional update of `ContextT`, then the doc should be tweaked I guess, e.g.:

```suggestion
 * injective function `f: (ContextT, SourceT) -> (ContextT, TargetT)`, that is, a function which, given a value of type
 * `SourceT`, computes a value of type `TargetT`, while generating an updated context of type `ContextT`.
 *
 * [[Cache]] defines an entry-point `getOrCreate`, such that
 *   - `getOrCreate(c, x) = f(c',x)`, if this is the first time `x` appears as the source value for `getOrCreate`.
 *   - `getOrCreate(c, x) = (c, v)`, if `getOrCreate(c', x)` was previously called, and returned `(c'', v)` (for some
 *     `c', c''`).
```